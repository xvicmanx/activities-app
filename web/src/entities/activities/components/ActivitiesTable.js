import moment from 'moment';
import React from 'react';
import CRUDTable, {
  CreateForm,
  DeleteForm,
  Fields,
  Field,
  Pagination,
  UpdateForm,
} from 'react-crud-table';

import { readTokenFromCookie } from '../../users/redux/UsersActions';
import ActivitiesService from '../services/ActivitiesService';
import CommunitiesDropdown from '../../communities/components/CommunitiesDropdown';

const styles = {
  container: {
    margin: 'auto',
    width: 'fit-content',
  },
};

const DescriptionRenderer = ({ field }) => <textarea {...field} />;
const CommunitiesSelectRenderer = ({ field }) => (
  <CommunitiesDropdown
    name={field.name}
    value={field.value}
    onChange={field.onChange}
  />
);

const service = {
  fetchItems: async () => {
    const response = await ActivitiesService.fetchActivities(
      readTokenFromCookie()
    );
    return response.activities;
  },
  fetchTotal: async () => {
    const response = await ActivitiesService.fetchActivities(
      readTokenFromCookie()
    );
    return response.activities.length;
  },
  create: (activity) =>
    ActivitiesService.createActivity(activity, readTokenFromCookie()),
  update: (activity) =>
    ActivitiesService.updateActivity(activity, readTokenFromCookie()),
  delete: (activity) =>
    ActivitiesService.deleteActivity(activity.id, readTokenFromCookie()),
};

const ActivitiesTable = () => (
  <div style={styles.container}>
    <CRUDTable
      caption="Actividades"
      fetchItems={(payload) => service.fetchItems(payload)}
      actionsLabel="Acciones"
    >
      <Fields>
        <Field name="id" label="Id" hideInCreateForm readOnly />
        <Field name="title" label="Título" placeholder="Título" />
        <Field
          name="description"
          label="Descripción"
          render={DescriptionRenderer}
        />
        <Field
          name="date"
          type="date"
          label="Fecha"
          tableValueResolver={(item) =>
            moment(item.date).format('DD-MM-YYYY HH:mm')
          }
        />
        <Field
          name="communityId"
          label="Comunidad"
          render={CommunitiesSelectRenderer}
          hideInUpdateForm
          hideFromTable
        />
      </Fields>
      <CreateForm
        title="Crear actividad"
        message="Crear una nueva actividad"
        trigger="Crear actividad"
        onSubmit={service.create}
        submitText="Crear"
        validate={(values) => {
          const errors = {};

          if (!values.title) {
            errors.title = 'Por favor, provea un título';
          }

          if (!values.description) {
            errors.description = 'Por favor, provea una Descripción';
          }

          if (!values.date) {
            errors.date = 'Por favor, provea una fecha';
          }

          if (!values.communityId) {
            errors.communityId = 'Por favor, seleccione una comunidad';
          }

          return errors;
        }}
      />
      <UpdateForm
        title="Actualizar actividad"
        message="Actualizar actividad"
        trigger="Actualizar"
        onSubmit={service.update}
        submitText="Actualizar"
        validate={(values) => {
          const errors = {};

          if (!values.id) {
            errors.id = 'Por favor, provea el id';
          }

          if (!values.title) {
            errors.title = 'Por favor, provea un título';
          }

          if (!values.description) {
            errors.description = 'Por favor, provea una Descripción';
          }

          if (!values.date) {
            errors.date = 'Por favor, provea una fecha';
          }

          return errors;
        }}
      />
      <DeleteForm
        title="Eliminar actividad"
        message="Esta seguro que quiere eliminar la actividad?"
        trigger="Eliminar"
        onSubmit={service.delete}
        submitText="Eliminar"
        validate={(values) => {
          const errors = {};

          if (!values.id) {
            errors.id = 'Por favor, provea el id';
          }

          return errors;
        }}
      />
      <Pagination
        itemsPerPage={10}
        fetchTotalOfItems={(payload) => service.fetchTotal(payload)}
      />
    </CRUDTable>
  </div>
);

export default ActivitiesTable;
