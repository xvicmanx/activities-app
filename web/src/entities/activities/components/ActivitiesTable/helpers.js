// @flow

import moment from 'moment';

export const dateValueTableValueResolver = (item: Object): any => moment(item.date).format('DD-MM-YYYY HH:mm');

