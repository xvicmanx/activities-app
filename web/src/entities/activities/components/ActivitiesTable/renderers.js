// @flow

import React from 'react';

import CommunitiesDropdown from '../../../communities/components/CommunitiesDropdown';


type RendererProps = {
  field: {
    value: string | number,
    name: string,
    onChange: Function,
  },
};

export const DescriptionRenderer = ({ field }: RendererProps): React$Element<any> => (
  <textarea {...field} />
);

export const CommunitiesSelectRenderer = ({ field }: RendererProps): React$Element<any> => (
  <CommunitiesDropdown
    name={field.name}
    value={field.value}
    onChange={field.onChange}
  />
);
