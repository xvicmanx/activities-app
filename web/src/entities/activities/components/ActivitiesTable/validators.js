export const deleteFormValidator = (values) => {
  const errors = {};

  if (!values.id) {
    errors.id = 'Por favor, provea el id';
  }

  return errors;
};

export const updateFormValidator = (values) => {
  const errors = {};

  if (!values.id) {
    errors.id = 'Por favor, provea el id';
  }

  if (!values.title) {
    errors.title = 'Por favor, provea un título';
  }

  if (!values.description) {
    errors.description = 'Por favor, provea una Descripción';
  }

  if (!values.date) {
    errors.date = 'Por favor, provea una fecha';
  }

  return errors;
};


export const createFormValidator = (values) => {
  const errors = {};

  if (!values.title) {
    errors.title = 'Por favor, provea un título';
  }

  if (!values.description) {
    errors.description = 'Por favor, provea una Descripción';
  }

  if (!values.date) {
    errors.date = 'Por favor, provea una fecha';
  }

  if (!values.communityId) {
    errors.communityId = 'Por favor, seleccione una comunidad';
  }

  return errors;
};
