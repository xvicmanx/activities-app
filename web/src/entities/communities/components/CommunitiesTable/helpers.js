// @flow

import React from 'react';
import { Link } from 'react-router-dom';

export const nameTableValueResolver = (item: Object): React$Element<any> => (
  <Link to={`/communities/${item.id}`}>{item.name}</Link>
);
