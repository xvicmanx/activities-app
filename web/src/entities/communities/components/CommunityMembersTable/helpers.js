// @flow

export const coordinatesTableValueResolver = (item: Object): string =>
  item.coordinates ? 'Si' : 'No';
