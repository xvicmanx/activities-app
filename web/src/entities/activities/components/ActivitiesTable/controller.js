// @flow

import { readTokenFromCookie } from '../../../users/redux/UsersActions';
import { encode } from '../../../../core/helpers';
import type { Options } from '../../../../core/helpers';
import ActivitiesService from '../../services/ActivitiesService';


const Controller = {
  fetchItems: async (options: Options): Promise<{items: any, total: any,...}> => {
    const response = await ActivitiesService.fetchActivities(
      readTokenFromCookie(),
      encode(options)
    );
    return {
      items: response.activities,
      total: response.total,
    };
  },
  create: async (activity: Object): Promise<any> => {
    try {
      const result = await ActivitiesService.createActivity(activity, readTokenFromCookie());
      return result;
    } catch (error) {
      return {
        success: false,
        message: error.message,
      };
    }
  },
  update: async (activity: Object): Promise<any> => {
    try {
      const result = await ActivitiesService.updateActivity(activity, readTokenFromCookie());
      return result;
    } catch (error) {
      return {
        success: false,
        message: error.message,
      };
    }
  },
  delete: async (activity: Object): Promise<any> => {
    try {
      const result = await ActivitiesService.deleteActivity(activity.id, readTokenFromCookie());
      return result;
    } catch (error) {
      return {
        success: false,
        message: error.message,
      };
    }
  }
};

export default Controller;
