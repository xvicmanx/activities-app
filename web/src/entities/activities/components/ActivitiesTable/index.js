// @flow

import React, { useState } from 'react';
import CRUDTable, {
  CreateForm,
  DeleteForm,
  Fields,
  Field,
  Pagination,
  UpdateForm,
} from 'react-crud-table';

import {
  DEFAULT_OPTIONS,
  throwErrorWhenNotSuccess,
} from '../../../../core/helpers';
import useTableItemsFetch from '../../../../core/useTableItemsFetch';

import {
  createFormValidator,
  deleteFormValidator,
  updateFormValidator,
} from './validators';
import { CommunitiesSelectRenderer, DescriptionRenderer } from './renderers';
import Controller from './controller';
import { dateValueTableValueResolver } from './helpers';

const ActivitiesTable = (): React$Element<any> => {
  const [options, setOptions] = useState(DEFAULT_OPTIONS);
  const result = useTableItemsFetch(
    Controller.fetchItems,
    options,
  );

  return (
    <CRUDTable
      caption="Actividades"
      items={result.items}
      actionsLabel="Acciones"
      showQueryBuilder
      onChange={(data) => {
        setOptions({
          sort: data.sort,
          queryRules: data.queryRules,
          activePage: data.pagination.activePage,
          itemsPerPage: data.pagination.itemsPerPage,
        });
      }}
    >
      <Fields>
        <Field name="id" label="Id" hideInCreateForm readOnly />
        <Field name="title" label="Título" placeholder="Título" />
        <Field
          name="description"
          label="Descripción"
          render={DescriptionRenderer}
        />
        <Field
          name="date"
          type="date"
          label="Fecha"
          tableValueResolver={dateValueTableValueResolver}
        />
        <Field
          name="communityId"
          label="Comunidad"
          render={CommunitiesSelectRenderer}
          hideInUpdateForm
          hideFromTable
        />
      </Fields>
      <CreateForm
        title="Crear actividad"
        message="Crear una nueva actividad"
        trigger="Crear actividad"
        onSubmit={throwErrorWhenNotSuccess(Controller.create)}
        submitText="Crear"
        validate={createFormValidator}
      />
      <UpdateForm
        title="Actualizar actividad"
        message="Actualizar actividad"
        trigger="Actualizar"
        onSubmit={throwErrorWhenNotSuccess(Controller.update)}
        submitText="Actualizar"
        validate={updateFormValidator}
      />
      <DeleteForm
        title="Eliminar actividad"
        message="Esta seguro que quiere eliminar la actividad?"
        trigger="Eliminar"
        onSubmit={throwErrorWhenNotSuccess(Controller.delete)}
        submitText="Eliminar"
        validate={deleteFormValidator}
      />
      <Pagination
        defaultActivePage={1}
        itemsPerPage={options.itemsPerPage}
        totalOfItems={result.total}
      />
    </CRUDTable>
  );
};

export default ActivitiesTable;
