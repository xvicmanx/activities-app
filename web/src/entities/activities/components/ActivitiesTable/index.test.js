import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { BrowserRouter as Router } from 'react-router-dom';
import { act } from 'react-dom/test-utils';

import { DEFAULT_OPTIONS } from '../../../../core/helpers';
import ActivitiesTable from './';
import Controller from './controller';

jest.mock('./renderers');
jest.mock('./renderers', () => ({
  DescriptionRenderer: () => 'DescriptionRenderer',
  CommunitiesSelectRenderer: () => 'CommunitiesSelectRenderer',
}));
jest.mock('../../../../core/useTableItemsFetch');

import useTableItemsFetch from '../../../../core/useTableItemsFetch';

Enzyme.configure({ adapter: new Adapter() });

describe('ActivitiesTable', () => {
  const activities = [
    {
      id: 1,
      title: 'Test Activity 1',
    },
    {
      id: 2,
      title: 'Test Activity 2',
    },
  ];
  let itemsResolver;

  beforeEach(() => {
    useTableItemsFetch.mockClear().mockImplementation(() => ({
      total: activities.length,
      items: activities,
    }));
  });

  it('renders properly', async () => {
    const result = Enzyme.mount(
      <Router>
        <ActivitiesTable />
      </Router>
    );

    expect(useTableItemsFetch).toHaveBeenCalledTimes(1);
    expect(useTableItemsFetch).toHaveBeenCalledWith(
      Controller.fetchItems,
      DEFAULT_OPTIONS
    );
    expect(result.find('CRUDTable').props()).toMatchSnapshot();
  });

  it('updates the options on change', async () => {
    const result = Enzyme.mount(
      <Router>
        <ActivitiesTable />
      </Router>
    );
    const data = {
      sort: {
        field: 'name',
        direction: 'descending',
      },
      queryRules: [],
      pagination: {
        activePage: 2,
        itemsPerPage: 15,
      },
    };

    result.find('CRUDTable').prop('onChange')(data);
    result.update();
    expect(result.find('CRUDTable').props()).toMatchSnapshot();
  });
});
