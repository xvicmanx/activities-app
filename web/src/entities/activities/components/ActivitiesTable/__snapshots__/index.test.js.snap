// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivitiesTable renders properly 1`] = `
Object {
  "actionsLabel": "Acciones",
  "caption": "Actividades",
  "children": Array [
    <CRUDTable_Fields>
      <CRUDTable_Field
        hideFromTable={false}
        hideInCreateForm={true}
        hideInUpdateForm={false}
        label="Id"
        name="id"
        queryable={true}
        readOnly={true}
        sortable={true}
        tableValueResolver={null}
        type="text"
      />
      <CRUDTable_Field
        hideFromTable={false}
        hideInCreateForm={false}
        hideInUpdateForm={false}
        label="Título"
        name="title"
        placeholder="Título"
        queryable={true}
        sortable={true}
        tableValueResolver={null}
        type="text"
      />
      <CRUDTable_Field
        hideFromTable={false}
        hideInCreateForm={false}
        hideInUpdateForm={false}
        label="Descripción"
        name="description"
        queryable={true}
        render={[Function]}
        sortable={true}
        tableValueResolver={null}
        type="text"
      />
      <CRUDTable_Field
        hideFromTable={false}
        hideInCreateForm={false}
        hideInUpdateForm={false}
        label="Fecha"
        name="date"
        queryable={true}
        sortable={true}
        tableValueResolver={[Function]}
        type="date"
      />
      <CRUDTable_Field
        hideFromTable={true}
        hideInCreateForm={false}
        hideInUpdateForm={true}
        label="Comunidad"
        name="communityId"
        queryable={true}
        render={[Function]}
        sortable={true}
        tableValueResolver={null}
        type="text"
      />
    </CRUDTable_Fields>,
    <CRUDTable_CreateForm
      message="Crear una nueva actividad"
      onSubmit={[Function]}
      submitText="Crear"
      title="Crear actividad"
      trigger="Crear actividad"
      validate={[Function]}
    />,
    <CRUDTable_UpdateForm
      message="Actualizar actividad"
      onSubmit={[Function]}
      submitText="Actualizar"
      title="Actualizar actividad"
      trigger="Actualizar"
      validate={[Function]}
    />,
    <CRUDTable_DeleteForm
      message="Esta seguro que quiere eliminar la actividad?"
      onSubmit={[Function]}
      submitText="Eliminar"
      title="Eliminar actividad"
      trigger="Eliminar"
      validate={[Function]}
    />,
    <CRUDTable_Pagination
      defaultActivePage={1}
      itemsPerPage={20}
      totalOfItems={2}
    />,
  ],
  "fetchItems": null,
  "items": Array [
    Object {
      "id": 1,
      "title": "Test Activity 1",
    },
    Object {
      "id": 2,
      "title": "Test Activity 2",
    },
  ],
  "onChange": [Function],
  "showQueryBuilder": true,
}
`;

exports[`ActivitiesTable updates the options on change 1`] = `
Object {
  "actionsLabel": "Acciones",
  "caption": "Actividades",
  "children": Array [
    <CRUDTable_Fields>
      <CRUDTable_Field
        hideFromTable={false}
        hideInCreateForm={true}
        hideInUpdateForm={false}
        label="Id"
        name="id"
        queryable={true}
        readOnly={true}
        sortable={true}
        tableValueResolver={null}
        type="text"
      />
      <CRUDTable_Field
        hideFromTable={false}
        hideInCreateForm={false}
        hideInUpdateForm={false}
        label="Título"
        name="title"
        placeholder="Título"
        queryable={true}
        sortable={true}
        tableValueResolver={null}
        type="text"
      />
      <CRUDTable_Field
        hideFromTable={false}
        hideInCreateForm={false}
        hideInUpdateForm={false}
        label="Descripción"
        name="description"
        queryable={true}
        render={[Function]}
        sortable={true}
        tableValueResolver={null}
        type="text"
      />
      <CRUDTable_Field
        hideFromTable={false}
        hideInCreateForm={false}
        hideInUpdateForm={false}
        label="Fecha"
        name="date"
        queryable={true}
        sortable={true}
        tableValueResolver={[Function]}
        type="date"
      />
      <CRUDTable_Field
        hideFromTable={true}
        hideInCreateForm={false}
        hideInUpdateForm={true}
        label="Comunidad"
        name="communityId"
        queryable={true}
        render={[Function]}
        sortable={true}
        tableValueResolver={null}
        type="text"
      />
    </CRUDTable_Fields>,
    <CRUDTable_CreateForm
      message="Crear una nueva actividad"
      onSubmit={[Function]}
      submitText="Crear"
      title="Crear actividad"
      trigger="Crear actividad"
      validate={[Function]}
    />,
    <CRUDTable_UpdateForm
      message="Actualizar actividad"
      onSubmit={[Function]}
      submitText="Actualizar"
      title="Actualizar actividad"
      trigger="Actualizar"
      validate={[Function]}
    />,
    <CRUDTable_DeleteForm
      message="Esta seguro que quiere eliminar la actividad?"
      onSubmit={[Function]}
      submitText="Eliminar"
      title="Eliminar actividad"
      trigger="Eliminar"
      validate={[Function]}
    />,
    <CRUDTable_Pagination
      defaultActivePage={1}
      itemsPerPage={15}
      totalOfItems={2}
    />,
  ],
  "fetchItems": null,
  "items": Array [
    Object {
      "id": 1,
      "title": "Test Activity 1",
    },
    Object {
      "id": 2,
      "title": "Test Activity 2",
    },
  ],
  "onChange": [Function],
  "showQueryBuilder": true,
}
`;
